test/data_tests/utilities_test.py

```diff
@@ -0,0 +1,218 @@
+"""
+This file tests the utilities stored in cassiopeia/data/utilities.py
+"""
+
+import unittest
+from typing import Dict, Optional
+
+import numpy as np
+import pandas as pd
+
+from cassiopeia.data import utilities as data_utilities
+from cassiopeia.preprocess import utilities as preprocessing_utilities
+
+
+class TestDataUtilities(unittest.TestCase):
+    def setUp(self):
+
+        # this should obey PP for easy checking of ancestral states
+        self.character_matrix = pd.DataFrame.from_dict(
+            {
+                "node3": [1, 0, 0, 0, 0, 0, 0, 0],
+                "node7": [1, 1, 0, 0, 0, 0, 0, 0],
+                "node9": [1, 1, 1, 0, 0, 0, 0, 0],
+                "node11": [1, 1, 1, 1, 0, 0, 0, 0],
+                "node13": [1, 1, 1, 1, 1, 0, 0, 0],
+                "node15": [1, 1, 1, 1, 1, 1, 0, 0],
+                "node17": [1, 1, 1, 1, 1, 1, 1, 0],
+                "node18": [1, 1, 1, 1, 1, 1, 1, 1],
+                "node5": [2, 0, 0, 0, 0, 0, 0, 0],
+                "node6": [2, 2, 0, 0, 0, 0, 0, 0],
+            },
+            orient="index",
+        )
+
+        self.priors = {
+            0: {1: 0.5, 2: 0.5},
+            1: {1: 0.4, 2: 0.6},
+            2: {1: 1.0},
+            3: {1: 1.0},
+            4: {1: 1.0},
+            5: {1: 1.0},
+            6: {1: 1.0},
+            7: {1: 1.0},
+        }
+
+        # Test allele table
+        at_dict = {
+            "cellBC": ["cellA", "cellA", "cellA", "cellB", "cellC"],
+            "intBC": ["A", "B", "C", "A", "C"],
+            "r1": ["None", "ATC", "GGG", "ATA", "GAA"],
+            "r2": ["None", "AAA", "GAA", "TTT", "GAA"],
+            "r3": ["ATC", "TTT", "ATA", "ATA", "ATA"],
+            "UMI": [5, 10, 1, 30, 30],
+        }
+
+        self.allele_table = pd.DataFrame.from_dict(at_dict)
+        self.indel_to_prior = pd.DataFrame.from_dict(
+            {
+                "ATC": 0.5,
+                "GGG": 0.2,
+                "GAA": 0.1,
+                "AAA": 0.05,
+                "TTT": 0.05,
+                "ATA": 0.1,
+            },
+            orient="index",
+            columns=["freq"],
+        )
+
+    def test_bootstrap_character_matrices_no_priors(self):
+
+        random_state = np.random.RandomState(123431235)
```

Comment:
I like your style.

Reply:
Is this an inside joke I've been thinking about this all afternoon